05_classify_mutations_multimodal/nbconverted/plot_multimodal_results.py

```diff
@@ -0,0 +1,304 @@
+print(compare_df.training_data.unique())
+compare_df[compare_df.gene == 'TP53'].head(10)
+
+
+# In[11]:
+
+
+sns.set({'figure.figsize': (10, 6)})
+
+plot_df = pd.DataFrame()
+
+# plot mean performance over all genes in pilot experiment
+for ix, gene in enumerate(results_df.identifier.unique()):
+    
+    plot_gene_df = compare_df[(compare_df.gene == gene)].reset_index(drop=True)
+    
+    # get the best-performing data types from the single-omics and multi-omics models
+    max_single_data_type = (
+        plot_gene_df[plot_gene_df.model_type == 'single-omics']
+          .groupby('training_data')
+          .agg('mean')
+          .delta_aupr.idxmax()
+    )
+    max_multi_data_type = (
+        plot_gene_df[plot_gene_df.model_type == 'multi-omics']
+          .groupby('training_data')
+          .agg('mean')
+          .delta_aupr.idxmax()
+    )
+    
+    # get samples with that data type
+    max_single_df = plot_gene_df[plot_gene_df.training_data == max_single_data_type]
+    max_multi_df = plot_gene_df[plot_gene_df.training_data == max_multi_data_type]
+    
+    # calculate difference between means and t-test p-val for that data type
+    mean_diff = max_single_df.delta_aupr.mean() - max_multi_df.delta_aupr.mean()
+    _, p_val = ttest_ind(max_single_df.delta_aupr.values,
+                         max_multi_df.delta_aupr.values)
+    print('{} diff: {:.4f} (pval: {:.4f})'.format(gene, mean_diff, p_val))
+    
+    plot_df = pd.concat((plot_df, max_single_df, max_multi_df))
+
+sns.boxplot(data=plot_df, x='gene', y='delta_aupr', hue='model_type')
+plt.title('Best performing single-omics vs. multi-omics data type, per gene')
+plt.xlabel('Target gene')
+plt.ylabel('AUPR(signal) - AUPR(shuffled)')
+
+
+# Summary and interpretation:
+# 
+# * In this last figure, we can see that the difference in distributions between the best-performing single-omics and multi-omics model is visually quite small for all of the target genes, and none of the differences are statistically significant using a t-test over the 8 total CV folds.
+# * Looking at the other two figures, when we consider all data types the difference between single-omics and multi-omics models is still pretty minimal in general. There are some genes where adding expression data helps a lot (EGFR, PIK3CA, KRAS) and others where it doesn't really (IDH1, SETD2), but in general there's not a lot of added predictive ability when expression is combined with other data types, suggesting that methylation may be most useful in that it correlates with expression patterns (although more follow-up would be needed to really confirm this).
+# * We had hypothesized that the reason for this might be that the multi-omics models are only selecting features from the best performing data type, but looking at the model coefficients this doesn't really seem to be the case; non-zero coefficients seem to be roughly evenly distributed between data types.
```

Comment:
I'm not sure if having "non-zero coefficients seem to be roughly evenly distributed between data types" is evidence enough that each data type is being evenly used. It could be that the features LASSO is choosing are features that are highly correlated between the two data types. (unless you are doing some PCA/ICA on the whole dataset first and only using those features, then look at the loadings to see if the most informative dimensions are driven by methylation or expression).

What would be cool here would be if you could somehow breakdown the proportion of variance explained between the data types to really show how much more information each data type is adding. This is a bit tricky and I am not sure what measure is best, but I could imagine this being done by looking at model fit and seeing how it changes with adding more variables (maybe AIC? or likelihood ratio test?).

Reply:
This is a good point about the coefficients/feature correlations. What you said here:

> if you could somehow breakdown the proportion of variance explained between the data types to really show how much more information each data type is adding

is exactly what I'd like to do, but I agree that it seems tricky. We'll definitely think about ways to do it going forward (keeping your ideas in mind), although I'm not sure if it will make the cut for this paper.

Doing PCA or other dimension reduction on the whole dataset was also something we thought about, but we decided not to include it in this analysis/paper, at least for now. We really don't want to spend a ton of time optimizing this process at this point - our main goal for this analysis isn't necessarily to show that fitting a good multimodal model for this problem isn't _possible_, more just to show that it's not trivially easy.